### PR TITLE
K8s 1.30

### PR DESCRIPTION
**Pull request by vardhaman22**
## Issue: 

 https://github.com/rancher/rancher/issues/45089
## Problem

 As per the issue k8s minor version 1.30 support needs to be added
## Solution

 updated all the k8s libraries to kubernetes v1.30.0
 
 other changes done:
 1. updated wrangler from v2 to v3
 
 2. replaced **k8s.io/gengo/examples/set-gen/sets** with "k8s.io/apimachinery/pkg/util/sets" since gengo is updated to gengo/v2 in kubernetes v1.30.0 and sets package is removed from gengo/v2 [ref](https://github.com/kubernetes/gengo/commit/f59fc19f8947553b5899bb065e15b692c38e8135#diff-6a811c40c584320a1942b76b03c23c4b7067d04f30c2555b962d30d19eecaf63)

 3. and since gengo args package is also removed in gengo/v2 so removed args.DefaultSourceTree() function uses and added the implementation of that function directly in the [code](https://github.com/rancher/rancher/pull/45483/commits/fe9a50e47c28adb8540b6a4ba02766689f74f89b). 
 
similar thing is done in a [PR](https://github.com/rancher/norman/pull/499/files#diff-ebe0c9061541da82c4de858d4996649ee6d33bf4036cb96200cc3f3ae891a536) for rancher/norman
 
 4. apiserver types moved from one dir to another
    https://github.com/kubernetes/apiserver/commit/e7eedd15ec9dd1ab5349b05ddbb1530655a28388#diff-179ab44de4a16de69e23c8a3fee5a0d6ef22755b90c182744f05ca9e1eca4381
    so updated imports in rancher [ref](https://github.com/rancher/rancher/pull/45483/commits/6aa9139ffe1580bac9d4245d707bf0a2cf0d375b)
 
5. bumped controller-runtime to v0.18.2 and cluster-api to v1.7.3 also updated capi crds and capi webhook yaml files.
updated capi webhooks with custom changes also which is rancher specific change [ref](https://github.com/rancher/rancher/pull/45483/commits/f5b2d7c7034d0bbde225d522819f122018406661)
even though capi v1.7.3 uses controller-runtime v0.17.3, but getting this error if controller-runtime is kept at v0.17.3
i am getting the following error so updated it to v0.18.4
```
# sigs.k8s.io/controller-runtime/pkg/metrics
vendor/sigs.k8s.io/controller-runtime/pkg/metrics/leaderelection.go:26:71: undefined: leaderelection.SwitchMetric
```

6. after updating capi crds an error occurred while running rancher
```
2024/05/17 08:06:42 [ERROR] error syncing 'cluster-member': handler auth-prov-v2-roletemplate: no matches for kind "Machine" in version "cluster.x-k8s.io/v1alpha3", requeuing
    2024/05/17 08:06:42 [ERROR] error syncing 'cluster-admin': handler auth-prov-v2-roletemplate: no matches for kind "Cluster" in version "cluster.x-k8s.io/v1alpha3", requeuing
    2024/05/17 08:06:43 [ERROR] error syncing 'cluster-owner': handler auth-prov-v2-roletemplate: no matches for kind "Cluster" in version "cluster.x-k8s.io/v1alpha3", requeuing
```
 cluster.x-k8s.io/v1alpha3 is deprecated in capi v1.7.2 but it is still in crd. So fixed rancher code to not add deprecated version of a resource from a crd to validate role data. [ref](https://github.com/rancher/rancher/pull/45483/commits/c9051713f04bdb6869b61c1f16a013a1abef14c3)
 
 7. updated python tests **test_node_fields and test_dns_fields**  fix needed because of addition of new fields in k8s v1.30 for these objects.
 
 8. updated local cluster k3s version to v1.30.1+k3s1 and also updated rancher chart kube-version annotation.
 9. bumped kubernetes libraries to v0.30.1
 10. updated golangci lint to v1.59.0.
    
### Manual Testing

1. tested v1.30.0 provisioning for all types of cluster.
Screenshot from 2024-05-31 21-40-54

2. tested upgrade from 1.29 to 1.30 for all types of cluster.
Screenshot from 2024-05-31 22-35-33
Screenshot from 2024-05-31 22-36-41
Screenshot from 2024-05-31 22-42-43

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

**Comment by mjura** [comment]
it has to be rebased

**Comment by slickwarren** [comment]
[shepherd PR](https://github.com/rancher/shepherd/pull/219) has been merged - ready for update of shepherd version in rancher/rancher

File: pkg/crds/provisioningv2/capi-webhooks.yaml [hunk 1]
```diff
@@ -2,479 +2,477 @@
...
+    resources:
+    - machinepools
+  sideEffects: None
+- admissionReviewVersions:
+  - v1
+  - v1beta1
+  clientConfig:
+    service:
+      name: capi-webhook-service
+      namespace: cattle-provisioning-capi-cattle-provisioning-capi-system
```
Comment (jiaqiluo) [inline pkg/crds/provisioningv2/capi-webhooks.yaml:189]:
```suggestion
      namespace: cattle-provisioning-capi-system
```

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/crds/provisioningv2/capi-webhooks.yaml (hunk 1 above)
Comment (snasovich) [reply pkg/crds/provisioningv2/capi-webhooks.yaml:189]: Is this file updated manually? A little concerning how this typo made it in.

Comment (vardhaman22) [reply pkg/crds/provisioningv2/capi-webhooks.yaml]: thanks @jiaqiluo for pointing it out.
yes @snasovich  i updated it manually ...I did two separate commits for it. First one just copies everything from upstream and then added a separate commit to update the namespace and service in the capi webhooks. When i faced issues with the upstream CRDs i checked earlier [commit](https://github.com/rancher/rancher/pull/42228/commits/8f2638e13f86e967033d2f90b0c9272c9b998934#diff-4bca09f24919c5585977ac50a0d2b82ab7e97ab1d1302326c391776fe3b959b9) in the repo and found out about this change. 
I looked at it now again and found that there is a conversion section added as well for a few crds so i updated them as well.

Comment (vardhaman22) [reply pkg/crds/provisioningv2/capi-webhooks.yaml]: had a discussion with @Oats87 and found out that these can be generated from a script in rancher/provisioning repo.
updating those files with the generated ones

**Review by jiaqiluo — APPROVED** [review]

**Review by kinarashah — APPROVED** [review]

**Pull request merged**